diagnostics: exclude indirect private deps from trait impl suggest

notriddle opened the pull request:
Fixes #88696

notriddle committed:
bootstrap: enable Cargo `public-dependency` feature for `libstd`

notriddle committed:
rustc_metadata: inherit dependency privacy flag

notriddle committed:
diagnostics: exclude indirect private deps from trait impl suggest

Fixes #88696

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4854132848/jobs/8651115445) [(plain)](https://github.com/rust-lang/rust/commit/84538188a292381dabe5a75eeb2b0bd59bacc6f2/checks/13152757516/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4854240468/jobs/8651363918) [(plain)](https://github.com/rust-lang/rust/commit/8bf782facd5355355d4968e899d9090975a09d10/checks/13153082174/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

cjgillot commented on compiler/rustc_metadata/src/creader.rs:
```diff
@@ -558,10 +560,11 @@ impl<'a, 'tcx> CrateLoader<'a, 'tcx> {
                     dep_kind = CrateDepKind::MacrosOnly;
                 }
                 data.update_dep_kind(|data_dep_kind| cmp::max(data_dep_kind, dep_kind));
+                data.update_private_dep(|private_dep| private_dep && private_dep);
```
This line does nothing. Did you mean this?

```suggestion
                data.update_private_dep(|p_d| p_d && private_dep);
```

notriddle committed:
Update compiler/rustc_metadata/src/creader.rs

Co-authored-by: Camille Gillot <gillot.camille@gmail.com>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4854663715/jobs/8652307866) [(plain)](https://github.com/rust-lang/rust/commit/a7efa7837f1c6d4557a4599a7cad3e76fdfaff86/checks/13154303260/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 14.18µs

 finished in 8.878 seconds
test: basic_install
$ sh /checkout/src/tools/rust-installer/gen-installer.sh --image-dir=/checkout/src/tools/rust-installer/test/image1 --work-dir=/checkout/obj/build/x86_64-unknown-linux-gnu/test/rust-installer/workdir --output-dir=/checkout/obj/build/x86_64-unknown-linux-gnu/test/rust-installer/outdir
error: failed to load manifest for workspace member `/checkout/library/std`
Caused by:
  failed to parse manifest at `/checkout/library/std/Cargo.toml`

Caused by:
Caused by:
  the cargo feature `public-dependency` requires a nightly version of Cargo, but this is the `beta` channel
  See https://doc.rust-lang.org/book/appendix-07-nightly-rust.html for more information about Rust release channels.
  See https://doc.rust-lang.org/beta/cargo/reference/unstable.html#public-dependency for more information about using this feature.
TEST FAILED!

Build completed unsuccessfully in 0:26:36
```

</details>

notriddle committed:
rust-installer: include `RUSTC_BOOTSTRAP` when generating installer

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4854907283/jobs/8652841842) [(plain)](https://github.com/rust-lang/rust/commit/f1d254f67858b3d53b71ed9067f3b72cda4d65ea/checks/13154994921/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test: multiple_components
test: uninstall_from_installed_script
test: uninstall_from_installed_script_with_args_fails
test: combine_installers
$ sh /checkout/src/tools/rust-installer/combine-installers.sh --work-dir=/checkout/obj/build/x86_64-unknown-linux-gnu/test/rust-installer/workdir --output-dir=/checkout/obj/build/x86_64-unknown-linux-gnu/test/rust-installer/outdir --package-name=rust --input-tarballs=/checkout/obj/build/x86_64-unknown-linux-gnu/test/rust-installer/outdir/rustc.tar.gz,/checkout/obj/build/x86_64-unknown-linux-gnu/test/rust-installer/outdir/cargo.tar.gz
error: failed to load manifest for workspace member `/checkout/library/std`
Caused by:
  failed to parse manifest at `/checkout/library/std/Cargo.toml`

Caused by:
Caused by:
  the cargo feature `public-dependency` requires a nightly version of Cargo, but this is the `beta` channel
  See https://doc.rust-lang.org/book/appendix-07-nightly-rust.html for more information about Rust release channels.
  See https://doc.rust-lang.org/beta/cargo/reference/unstable.html#public-dependency for more information about using this feature.
TEST FAILED!

Build completed unsuccessfully in 0:30:47
```

</details>

b-naber commented on compiler/rustc_metadata/src/rmeta/decoder.rs:
```diff
@@ -111,7 +111,7 @@ pub(crate) struct CrateMetadata {
     /// Whether or not this crate should be consider a private dependency
     /// for purposes of the 'exported_private_dependencies' lint
-    private_dep: bool,
+    private_dep: Lock<bool>,
```
`AtomicBool` should suffice here.

b-naber reviewed (commented):
The approach seems better than the previous one, but I'm not sure I fully understand what you're doing here since I don't really know the code in `rustc_metadata`.

b-naber commented:
r? rust-lang/compiler

cjgillot commented on library/std/tests/common/mod.rs:
```diff
@@ -1,5 +1,7 @@
 #![allow(unused)]
 
+#![allow(exported_private_dependencies)]
```
Why is this needed?

bors commented:
:umbrella: The latest upstream changes (presumably #111346) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

notriddle committed:
Improve comments

notriddle committed:
rustc_metadata: fix private_dep logic in `register_crate`

notriddle committed:
diagnostics: don't crash if an injected crate shows up in suggestions

notriddle committed:
rustc_metadata: use AtomicBool for privateness instead of Lock

notriddle committed:
remove outdated comment from `is_user_visible_dep` docs

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4919860991/jobs/8787999240) [(plain)](https://github.com/rust-lang/rust/commit/97264e2349fa00649a5468b99ca5c31de99b7e98/checks/13326213457/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

notriddle committed:
std: mark common functions in test crate `pub(crate)`

This is not a library, so there's no reason for them to be `pub`.
Without doing this, compiling the test crates causes private dep
lint errors:

      error: type `PathBuf` from private dependency 'std' in public interface
        --> library/std/tests/common/mod.rs:26:5
         |
      26 |     pub fn join(&self, path: &str) -> PathBuf {
         |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
         |
         = note: `-D exported-private-dependencies` implied by `-D warnings`

      error: type `Path` from private dependency 'std' in public interface
        --> library/std/tests/common/mod.rs:31:5
         |
      31 |     pub fn path(&self) -> &Path {
         |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^

      error: could not compile `std` (test "create_dir_all_bare") due to 2 previous errors

This happens because Cargo passes `--extern 'priv:std=...` when
compiling the test crate.

I'm not sure if these warnings are desirable or not. They seem correct
in a very pedantic way (the dependency on `std` is not marked public,
since it's implicit), but also pointless (the test crate is not an API,
so who cares what it does).

klensy commented on compiler/rustc_metadata/src/rmeta/decoder.rs:
```diff
@@ -109,9 +110,10 @@ pub(crate) struct CrateMetadata {
+    /// Whether or not this crate should be consider a private dependency.
+    /// Used by the 'exported_private_dependencies' lint, and for determining
+    /// whether to emit suggestions that reference this crate.
+    private_dep: AtomicBool,
```
Or maybe use https://github.com/rust-lang/rust/blob/dfe31889e10e36eed53327d1ca624fbf21b475a5/compiler/rustc_data_structures/src/sync.rs#L24?

notriddle committed:
rustc_metadata: use configurable AtomicBool for privateness flag

This switches to using a `Cell` for single-threaded rustc.

cjgillot commented on compiler/rustc_middle/src/ty/util.rs:
```diff
@@ -808,6 +808,26 @@ impl<'tcx> TyCtxt<'tcx> {
+    pub fn is_user_visible_dep(self, key: CrateNum) -> bool {
+        // | Private | Direct | Visible |                    |
+        // |---------|--------|---------|--------------------|
+        // | Yes     | Yes    | Yes     | !(true && !true)   |
+        // | No      | Yes    | Yes     | !(false && !true)  |
+        // | Yes     | No     | No      | !(true && !false)  |
+        // | No      | No     | Yes     | !(false && !false) |
+        !(self.is_private_dep(key)
+            // If `extern_crate` is `None`, then the crate was injected (e.g., by the allocator).
+            // Treat that kind of crate as "indirect", since it's an implementation detail of
+            // the language.
+            && !self.extern_crate(key.as_def_id()).map_or(false, |e| e.is_direct()))
```
```suggestion
        // | Yes     | Yes    | Yes     | !true || true   |
        // | No      | Yes    | Yes     | !false || true  |
        // | Yes     | No     | No      | !true || false  |
        // | No      | No     | Yes     | !false || false |
        !self.is_private_dep(key)
            // If `extern_crate` is `None`, then the crate was injected (e.g., by the allocator).
            // Treat that kind of crate as "indirect", since it's an implementation detail of
            // the language.
            || self.extern_crate(key.as_def_id()).map_or(false, |e| e.is_direct()))
```

cjgillot reviewed (commented):
I only have 2 nits. I'd like to get some input from the bootstrap team, as this PR introduces RUSTC_BOOTSTRAP=1 in a few places. cc @Mark-Simulacrum (or please reassign)

notriddle committed:
Use De Morgan's law to simplify logic

Co-authored-by: Camille Gillot <gillot.camille@gmail.com>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4930185763/jobs/8810783028) [(plain)](https://github.com/rust-lang/rust/commit/1c7f14bff06d0cff6ba765993193649746f37dfc/checks/13355465514/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_middle v0.0.0 (/checkout/compiler/rustc_middle)
error: mismatched closing delimiter: `)`
   --> compiler/rustc_middle/src/ty/util.rs:818:61
    |
818 |     pub fn is_user_visible_dep(self, key: CrateNum) -> bool {
...
...
829 |             || self.extern_crate(key.as_def_id()).map_or(false, |e| e.is_direct()))

error: unexpected closing delimiter: `}`
   --> compiler/rustc_middle/src/ty/util.rs:831:1
    |
    |
829 |             || self.extern_crate(key.as_def_id()).map_or(false, |e| e.is_direct()))
    |                                                                                   - missing open `(` for this delimiter
831 | }
    | ^ unexpected closing delimiter

error: could not compile `rustc_middle` (lib test) due to 2 previous errors
```

</details>

notriddle committed:
rustc_metadata: specialize private_dep flag with `fetch_and`

bors commented:
:pushpin: Commit bbc967400dcfe4be5a7fa7216fdf57699e76cd85 has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot bbc967400dcfe4be5a7fa7216fdf57699e76cd85 -->
<!-- homu: {"type":"Approved","sha":"bbc967400dcfe4be5a7fa7216fdf57699e76cd85","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit bbc967400dcfe4be5a7fa7216fdf57699e76cd85 with merge f284f7ecf5acf7413462a37eda17bcaca56be266...
<!-- homu: {"type":"BuildStarted","head_sha":"bbc967400dcfe4be5a7fa7216fdf57699e76cd85","merge_sha":"f284f7ecf5acf7413462a37eda17bcaca56be266"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4987705158/jobs/8929787424)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/4987705158/jobs/8929787424","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-distcheck`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/4987705158/jobs/8929719326) [(plain)](https://github.com/rust-lang-ci/rust/commit/f284f7ecf5acf7413462a37eda17bcaca56be266/checks/13505669244/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: 
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/obj/build/tmp/distcheck/x.py --help`
error: vendoring required, but vendor directory does not exist.
       Run `cargo vendor --sync ./src/tools/cargo/Cargo.toml --sync ./src/tools/rust-analyzer/Cargo.toml --sync ./compiler/rustc_codegen_cranelift/Cargo.toml --sync ./src/bootstrap/Cargo.toml ` to initialize the vendor directory.
Alternatively, use the pre-vendored `rustc-src` dist component.
  File "/checkout/obj/build/tmp/distcheck/src/bootstrap/bootstrap.py", line 1100, in <module>
    main()
  File "/checkout/obj/build/tmp/distcheck/src/bootstrap/bootstrap.py", line 1084, in main
    bootstrap(args)
    bootstrap(args)
  File "/checkout/obj/build/tmp/distcheck/src/bootstrap/bootstrap.py", line 1031, in bootstrap
    build.check_vendored_status()
  File "/checkout/obj/build/tmp/distcheck/src/bootstrap/bootstrap.py", line 965, in check_vendored_status
    raise Exception("{} not found".format(vendor_dir))
Exception: /checkout/obj/build/tmp/distcheck/vendor not found
make: *** [Makefile:42: check] Error 1
```

</details>

notriddle committed:
bootstrap: use RUSTC_BOOTSTRAP to vendor sources

klensy commented on src/bootstrap/dist.rs:
```diff
@@ -1013,6 +1013,7 @@ impl Step for PlainSourceTarball {
                 .arg(builder.src.join("./compiler/rustc_codegen_cranelift/Cargo.toml"))
                 .arg("--sync")
                 .arg(builder.src.join("./src/bootstrap/Cargo.toml"))
+                .env("RUSTC_BOOTSTRAP", "1")
```
Add similar comment as for other `RUSTC_BOOTSTRAP` here too?

cjgillot commented:
@bors r+

bors commented:
:pushpin: Commit b261f7850e0f3d2e7da91dbcef6670bbcdbad6ed has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot b261f7850e0f3d2e7da91dbcef6670bbcdbad6ed -->
<!-- homu: {"type":"Approved","sha":"b261f7850e0f3d2e7da91dbcef6670bbcdbad6ed","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit b261f7850e0f3d2e7da91dbcef6670bbcdbad6ed with merge bc198e5dc8b421b45ceea2d4a698a50317a21f04...
<!-- homu: {"type":"BuildStarted","head_sha":"b261f7850e0f3d2e7da91dbcef6670bbcdbad6ed","merge_sha":"bc198e5dc8b421b45ceea2d4a698a50317a21f04"} -->

rust-log-analyzer commented:
The job **`dist-various-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/5006484691/jobs/8971832219) [(plain)](https://github.com/rust-lang-ci/rust/commit/bc198e5dc8b421b45ceea2d4a698a50317a21f04/checks/13559931388/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] gimli test:false 3.433
[RUSTC-TIMING] object test:false 4.050
warning: dropping unsupported crate type `dylib` for target `wasm32-wasi`

error: type `Advice` from private dependency 'wasi' in public interface
  --> library/std/src/sys/wasi/fd.rs:99:5
   |
99 |     pub fn advise(&self, offset: u64, len: u64, advice: wasi::Advice) -> io::Result<()> {
   |
   |
   = note: `-D exported-private-dependencies` implied by `-D warnings`

error: type `Filestat` from private dependency 'wasi' in public interface
   --> library/std/src/sys/wasi/fd.rs:182:5
    |
182 |     pub fn filestat_get(&self) -> io::Result<wasi::Filestat> {


error: type `Filestat` from private dependency 'wasi' in public interface
   --> library/std/src/sys/wasi/fd.rs:202:5
202 | /     pub fn path_filestat_get(
203 | |         &self,
203 | |         &self,
204 | |         flags: wasi::Lookupflags,
205 | |         path: &str,
206 | |     ) -> io::Result<wasi::Filestat> {


error: type `Filestat` from private dependency 'wasi' in public interface
   --> library/std/src/sys/wasi/fs.rs:107:5
    |
107 |     pub fn as_wasi(&self) -> &wasi::Filestat {


error: type `Filetype` from private dependency 'wasi' in public interface
   --> library/std/src/sys/wasi/fs.rs:145:5
    |
145 |     pub fn bits(&self) -> wasi::Filetype {

[RUSTC-TIMING] std test:false 2.121
warning: `std` (lib) generated 1 warning
error: could not compile `std` (lib) due to 5 previous errors; 1 warning emitted
```

</details>

notriddle committed:
std: make internal-only items `pub(crate)`

This works around a weird problem that looks like a bug in the
`exported_private_dependencies` lint.

bors commented:
:pushpin: Commit 568c6846bb4bd1659425585977883f7a74cfe40e has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 568c6846bb4bd1659425585977883f7a74cfe40e -->
<!-- homu: {"type":"Approved","sha":"568c6846bb4bd1659425585977883f7a74cfe40e","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 568c6846bb4bd1659425585977883f7a74cfe40e with merge f45885282155b8c288089a75e8951417ac16c6ba...
<!-- homu: {"type":"BuildStarted","head_sha":"568c6846bb4bd1659425585977883f7a74cfe40e","merge_sha":"f45885282155b8c288089a75e8951417ac16c6ba"} -->

rust-log-analyzer commented:
The job **`dist-various-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/5028886449/jobs/9020021164) [(plain)](https://github.com/rust-lang-ci/rust/commit/f45885282155b8c288089a75e8951417ac16c6ba/checks/13622099683/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] gimli test:false 3.998
[RUSTC-TIMING] object test:false 4.504
warning: dropping unsupported crate type `dylib` for target `x86_64-fortanix-unknown-sgx`

error: type `fortanix_sgx_abi::ByteBuffer` from private dependency 'fortanix_sgx_abi' in public interface
   --> library/std/src/sys/sgx/abi/usercalls/alloc.rs:839:5
839 |     pub fn copy_user_buffer(&self) -> Vec<u8> {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    |
    = note: `-D exported-private-dependencies` implied by `-D warnings`

error: type `fortanix_sgx_abi::ByteBuffer` from private dependency 'fortanix_sgx_abi' in public interface
   --> library/std/src/sys/sgx/abi/usercalls/raw.rs:227:9
60  | / macro_rules! define_usercalls {
60  | / macro_rules! define_usercalls {
61  | |     ($(fn $f:ident($($n:ident: $t:ty),*) $(-> $r:tt)*; )*) => {
62  | |         /// Usercall numbers as per the ABI.
63  | |         #[repr(u64)]
...   |
74  | |         $(enclave_usercalls_internal_define_usercalls!(def fn $f($($n: $t),*) $(-> $r)*);)*
75  | |     };
76  | | }
    | |_- in this expansion of `define_usercalls!` (#2)
...
...
188 | / macro_rules! enclave_usercalls_internal_define_usercalls {
189 | |     (def fn $f:ident($n1:ident: $t1:ty, $n2:ident: $t2:ty,
190 | |                      $n3:ident: $t3:ty, $n4:ident: $t4:ty) -> $r:tt) => (
191 | |         /// This is the raw function definition, see the ABI documentation for
...   |
227 | |         pub unsafe fn $f($n1: $t1, $n2: $t2) -> $r {
...   |
266 | |     );
267 | | }
267 | | }
    | |_- in this expansion of `enclave_usercalls_internal_define_usercalls!` (#3)
268 |
269 |   invoke_with_usercalls!(define_usercalls);
    |
   ::: /cargo/registry/src/index.crates.io-6f17d22bba15001f/fortanix-sgx-abi-0.5.0/src/lib.rs:873:43
    |
    |
873 |           { define_invoke_with_usercalls!(@ [$($accumulated)* $(fn $f($($n: $t),*) $(-> $r)*;)*] $($remainder)*); };
874 | |     // visit modules
874 | |     // visit modules
875 | |     (@ $accumulated:tt $(#[$meta:meta])* pub mod $modname:ident { $($contents:tt)* } $($remainder:tt)*) =>
876 | |         { define_invoke_with_usercalls!(@ $accumulated $($contents)* $($remainder)*); };
892 | |         macro_rules! invoke_with_usercalls {
892 | |         macro_rules! invoke_with_usercalls {
    | |         ---------------------------------- in this expansion of `invoke_with_usercalls!` (#1)
893 | |             ($m:ident) => { $m! $accumulated; }


error: type `fortanix_sgx_abi::ByteBuffer` from private dependency 'fortanix_sgx_abi' in public interface
   --> library/std/src/sys/sgx/abi/usercalls/raw.rs:211:9
60  | / macro_rules! define_usercalls {
60  | / macro_rules! define_usercalls {
61  | |     ($(fn $f:ident($($n:ident: $t:ty),*) $(-> $r:tt)*; )*) => {
62  | |         /// Usercall numbers as per the ABI.
63  | |         #[repr(u64)]
...   |
74  | |         $(enclave_usercalls_internal_define_usercalls!(def fn $f($($n: $t),*) $(-> $r)*);)*
75  | |     };
76  | | }
    | |_- in this expansion of `define_usercalls!` (#2)
...
...
188 | / macro_rules! enclave_usercalls_internal_define_usercalls {
189 | |     (def fn $f:ident($n1:ident: $t1:ty, $n2:ident: $t2:ty,
190 | |                      $n3:ident: $t3:ty, $n4:ident: $t4:ty) -> $r:tt) => (
191 | |         /// This is the raw function definition, see the ABI documentation for
...   |
211 | |         pub unsafe fn $f($n1: $t1, $n2: $t2, $n3: $t3) -> $r {
...   |
266 | |     );
267 | | }
267 | | }
    | |_- in this expansion of `enclave_usercalls_internal_define_usercalls!` (#3)
268 |
269 |   invoke_with_usercalls!(define_usercalls);
    |
   ::: /cargo/registry/src/index.crates.io-6f17d22bba15001f/fortanix-sgx-abi-0.5.0/src/lib.rs:873:43
    |
    |
873 |           { define_invoke_with_usercalls!(@ [$($accumulated)* $(fn $f($($n: $t),*) $(-> $r)*;)*] $($remainder)*); };
874 | |     // visit modules
874 | |     // visit modules
875 | |     (@ $accumulated:tt $(#[$meta:meta])* pub mod $modname:ident { $($contents:tt)* } $($remainder:tt)*) =>
876 | |         { define_invoke_with_usercalls!(@ $accumulated $($contents)* $($remainder)*); };
892 | |         macro_rules! invoke_with_usercalls {
892 | |         macro_rules! invoke_with_usercalls {
    | |         ---------------------------------- in this expansion of `invoke_with_usercalls!` (#1)
893 | |             ($m:ident) => { $m! $accumulated; }


error: type `fortanix_sgx_abi::ByteBuffer` from private dependency 'fortanix_sgx_abi' in public interface
   --> library/std/src/sys/sgx/abi/usercalls/raw.rs:195:9
60  | / macro_rules! define_usercalls {
60  | / macro_rules! define_usercalls {
61  | |     ($(fn $f:ident($($n:ident: $t:ty),*) $(-> $r:tt)*; )*) => {
62  | |         /// Usercall numbers as per the ABI.
63  | |         #[repr(u64)]
...   |
74  | |         $(enclave_usercalls_internal_define_usercalls!(def fn $f($($n: $t),*) $(-> $r)*);)*
75  | |     };
76  | | }
    | |_- in this expansion of `define_usercalls!` (#2)
...
...
188 | / macro_rules! enclave_usercalls_internal_define_usercalls {
189 | |     (def fn $f:ident($n1:ident: $t1:ty, $n2:ident: $t2:ty,
190 | |                      $n3:ident: $t3:ty, $n4:ident: $t4:ty) -> $r:tt) => (
191 | |         /// This is the raw function definition, see the ABI documentation for
...   |
195 | |         pub unsafe fn $f($n1: $t1, $n2: $t2, $n3: $t3, $n4: $t4) -> $r {
...   |
266 | |     );
267 | | }
267 | | }
    | |_- in this expansion of `enclave_usercalls_internal_define_usercalls!` (#3)
268 |
269 |   invoke_with_usercalls!(define_usercalls);
    |
   ::: /cargo/registry/src/index.crates.io-6f17d22bba15001f/fortanix-sgx-abi-0.5.0/src/lib.rs:873:43
    |
    |
873 |           { define_invoke_with_usercalls!(@ [$($accumulated)* $(fn $f($($n: $t),*) $(-> $r)*;)*] $($remainder)*); };
874 | |     // visit modules
874 | |     // visit modules
875 | |     (@ $accumulated:tt $(#[$meta:meta])* pub mod $modname:ident { $($contents:tt)* } $($remainder:tt)*) =>
876 | |         { define_invoke_with_usercalls!(@ $accumulated $($contents)* $($remainder)*); };
892 | |         macro_rules! invoke_with_usercalls {
892 | |         macro_rules! invoke_with_usercalls {
    | |         ---------------------------------- in this expansion of `invoke_with_usercalls!` (#1)
893 | |             ($m:ident) => { $m! $accumulated; }


error: type `fortanix_sgx_abi::FifoDescriptor<fortanix_sgx_abi::Usercall>` from private dependency 'fortanix_sgx_abi' in public interface
   --> library/std/src/sys/sgx/abi/usercalls/raw.rs:211:9
60  | / macro_rules! define_usercalls {
60  | / macro_rules! define_usercalls {
61  | |     ($(fn $f:ident($($n:ident: $t:ty),*) $(-> $r:tt)*; )*) => {
62  | |         /// Usercall numbers as per the ABI.
63  | |         #[repr(u64)]
...   |
74  | |         $(enclave_usercalls_internal_define_usercalls!(def fn $f($($n: $t),*) $(-> $r)*);)*
75  | |     };
76  | | }
    | |_- in this expansion of `define_usercalls!` (#2)
...
...
188 | / macro_rules! enclave_usercalls_internal_define_usercalls {
189 | |     (def fn $f:ident($n1:ident: $t1:ty, $n2:ident: $t2:ty,
190 | |                      $n3:ident: $t3:ty, $n4:ident: $t4:ty) -> $r:tt) => (
191 | |         /// This is the raw function definition, see the ABI documentation for
...   |
211 | |         pub unsafe fn $f($n1: $t1, $n2: $t2, $n3: $t3) -> $r {
...   |
266 | |     );
267 | | }
267 | | }
    | |_- in this expansion of `enclave_usercalls_internal_define_usercalls!` (#3)
268 |
269 |   invoke_with_usercalls!(define_usercalls);
    |
   ::: /cargo/registry/src/index.crates.io-6f17d22bba15001f/fortanix-sgx-abi-0.5.0/src/lib.rs:873:43
    |
    |
873 |           { define_invoke_with_usercalls!(@ [$($accumulated)* $(fn $f($($n: $t),*) $(-> $r)*;)*] $($remainder)*); };
874 | |     // visit modules
874 | |     // visit modules
875 | |     (@ $accumulated:tt $(#[$meta:meta])* pub mod $modname:ident { $($contents:tt)* } $($remainder:tt)*) =>
876 | |         { define_invoke_with_usercalls!(@ $accumulated $($contents)* $($remainder)*); };
892 | |         macro_rules! invoke_with_usercalls {
892 | |         macro_rules! invoke_with_usercalls {
    | |         ---------------------------------- in this expansion of `invoke_with_usercalls!` (#1)
893 | |             ($m:ident) => { $m! $accumulated; }


error: type `fortanix_sgx_abi::Usercall` from private dependency 'fortanix_sgx_abi' in public interface
   --> library/std/src/sys/sgx/abi/usercalls/raw.rs:211:9
60  | / macro_rules! define_usercalls {
60  | / macro_rules! define_usercalls {
61  | |     ($(fn $f:ident($($n:ident: $t:ty),*) $(-> $r:tt)*; )*) => {
62  | |         /// Usercall numbers as per the ABI.
63  | |         #[repr(u64)]
...   |
74  | |         $(enclave_usercalls_internal_define_usercalls!(def fn $f($($n: $t),*) $(-> $r)*);)*
75  | |     };
76  | | }
    | |_- in this expansion of `define_usercalls!` (#2)
...
...
188 | / macro_rules! enclave_usercalls_internal_define_usercalls {
189 | |     (def fn $f:ident($n1:ident: $t1:ty, $n2:ident: $t2:ty,
190 | |                      $n3:ident: $t3:ty, $n4:ident: $t4:ty) -> $r:tt) => (
191 | |         /// This is the raw function definition, see the ABI documentation for
...   |
211 | |         pub unsafe fn $f($n1: $t1, $n2: $t2, $n3: $t3) -> $r {
...   |
266 | |     );
267 | | }
267 | | }
    | |_- in this expansion of `enclave_usercalls_internal_define_usercalls!` (#3)
268 |
269 |   invoke_with_usercalls!(define_usercalls);
    |
   ::: /cargo/registry/src/index.crates.io-6f17d22bba15001f/fortanix-sgx-abi-0.5.0/src/lib.rs:873:43
    |
    |
873 |           { define_invoke_with_usercalls!(@ [$($accumulated)* $(fn $f($($n: $t),*) $(-> $r)*;)*] $($remainder)*); };
874 | |     // visit modules
874 | |     // visit modules
875 | |     (@ $accumulated:tt $(#[$meta:meta])* pub mod $modname:ident { $($contents:tt)* } $($remainder:tt)*) =>
876 | |         { define_invoke_with_usercalls!(@ $accumulated $($contents)* $($remainder)*); };
892 | |         macro_rules! invoke_with_usercalls {
892 | |         macro_rules! invoke_with_usercalls {
    | |         ---------------------------------- in this expansion of `invoke_with_usercalls!` (#1)
893 | |             ($m:ident) => { $m! $accumulated; }


error: type `fortanix_sgx_abi::FifoDescriptor<fortanix_sgx_abi::Return>` from private dependency 'fortanix_sgx_abi' in public interface
   --> library/std/src/sys/sgx/abi/usercalls/raw.rs:211:9
60  | / macro_rules! define_usercalls {
60  | / macro_rules! define_usercalls {
61  | |     ($(fn $f:ident($($n:ident: $t:ty),*) $(-> $r:tt)*; )*) => {
62  | |         /// Usercall numbers as per the ABI.
63  | |         #[repr(u64)]
...   |
74  | |         $(enclave_usercalls_internal_define_usercalls!(def fn $f($($n: $t),*) $(-> $r)*);)*
75  | |     };
76  | | }
    | |_- in this expansion of `define_usercalls!` (#2)
...
...
188 | / macro_rules! enclave_usercalls_internal_define_usercalls {
189 | |     (def fn $f:ident($n1:ident: $t1:ty, $n2:ident: $t2:ty,
190 | |                      $n3:ident: $t3:ty, $n4:ident: $t4:ty) -> $r:tt) => (
191 | |         /// This is the raw function definition, see the ABI documentation for
...   |
211 | |         pub unsafe fn $f($n1: $t1, $n2: $t2, $n3: $t3) -> $r {
...   |
266 | |     );
267 | | }
267 | | }
    | |_- in this expansion of `enclave_usercalls_internal_define_usercalls!` (#3)
268 |
269 |   invoke_with_usercalls!(define_usercalls);
    |
   ::: /cargo/registry/src/index.crates.io-6f17d22bba15001f/fortanix-sgx-abi-0.5.0/src/lib.rs:873:43
    |
    |
873 |           { define_invoke_with_usercalls!(@ [$($accumulated)* $(fn $f($($n: $t),*) $(-> $r)*;)*] $($remainder)*); };
874 | |     // visit modules
874 | |     // visit modules
875 | |     (@ $accumulated:tt $(#[$meta:meta])* pub mod $modname:ident { $($contents:tt)* } $($remainder:tt)*) =>
876 | |         { define_invoke_with_usercalls!(@ $accumulated $($contents)* $($remainder)*); };
892 | |         macro_rules! invoke_with_usercalls {
892 | |         macro_rules! invoke_with_usercalls {
    | |         ---------------------------------- in this expansion of `invoke_with_usercalls!` (#1)
893 | |             ($m:ident) => { $m! $accumulated; }


error: type `fortanix_sgx_abi::Return` from private dependency 'fortanix_sgx_abi' in public interface
   --> library/std/src/sys/sgx/abi/usercalls/raw.rs:211:9
60  | / macro_rules! define_usercalls {
60  | / macro_rules! define_usercalls {
61  | |     ($(fn $f:ident($($n:ident: $t:ty),*) $(-> $r:tt)*; )*) => {
62  | |         /// Usercall numbers as per the ABI.
63  | |         #[repr(u64)]
...   |
74  | |         $(enclave_usercalls_internal_define_usercalls!(def fn $f($($n: $t),*) $(-> $r)*);)*
75  | |     };
76  | | }
    | |_- in this expansion of `define_usercalls!` (#2)
...
...
188 | / macro_rules! enclave_usercalls_internal_define_usercalls {
189 | |     (def fn $f:ident($n1:ident: $t1:ty, $n2:ident: $t2:ty,
190 | |                      $n3:ident: $t3:ty, $n4:ident: $t4:ty) -> $r:tt) => (
191 | |         /// This is the raw function definition, see the ABI documentation for
...   |
211 | |         pub unsafe fn $f($n1: $t1, $n2: $t2, $n3: $t3) -> $r {
...   |
266 | |     );
267 | | }
267 | | }
    | |_- in this expansion of `enclave_usercalls_internal_define_usercalls!` (#3)
268 |
269 |   invoke_with_usercalls!(define_usercalls);
    |
   ::: /cargo/registry/src/index.crates.io-6f17d22bba15001f/fortanix-sgx-abi-0.5.0/src/lib.rs:873:43
    |
    |
873 |           { define_invoke_with_usercalls!(@ [$($accumulated)* $(fn $f($($n: $t),*) $(-> $r)*;)*] $($remainder)*); };
874 | |     // visit modules
874 | |     // visit modules
875 | |     (@ $accumulated:tt $(#[$meta:meta])* pub mod $modname:ident { $($contents:tt)* } $($remainder:tt)*) =>
876 | |         { define_invoke_with_usercalls!(@ $accumulated $($contents)* $($remainder)*); };
892 | |         macro_rules! invoke_with_usercalls {
892 | |         macro_rules! invoke_with_usercalls {
    | |         ---------------------------------- in this expansion of `invoke_with_usercalls!` (#1)
893 | |             ($m:ident) => { $m! $accumulated; }


error: type `fortanix_sgx_abi::FifoDescriptor<fortanix_sgx_abi::Cancel>` from private dependency 'fortanix_sgx_abi' in public interface
   --> library/std/src/sys/sgx/abi/usercalls/raw.rs:211:9
60  | / macro_rules! define_usercalls {
60  | / macro_rules! define_usercalls {
61  | |     ($(fn $f:ident($($n:ident: $t:ty),*) $(-> $r:tt)*; )*) => {
62  | |         /// Usercall numbers as per the ABI.
63  | |         #[repr(u64)]
...   |
74  | |         $(enclave_usercalls_internal_define_usercalls!(def fn $f($($n: $t),*) $(-> $r)*);)*
75  | |     };
76  | | }
    | |_- in this expansion of `define_usercalls!` (#2)
...
...
188 | / macro_rules! enclave_usercalls_internal_define_usercalls {
189 | |     (def fn $f:ident($n1:ident: $t1:ty, $n2:ident: $t2:ty,
190 | |                      $n3:ident: $t3:ty, $n4:ident: $t4:ty) -> $r:tt) => (
191 | |         /// This is the raw function definition, see the ABI documentation for
...   |
211 | |         pub unsafe fn $f($n1: $t1, $n2: $t2, $n3: $t3) -> $r {
...   |
266 | |     );
267 | | }
267 | | }
    | |_- in this expansion of `enclave_usercalls_internal_define_usercalls!` (#3)
268 |
269 |   invoke_with_usercalls!(define_usercalls);
    |
   ::: /cargo/registry/src/index.crates.io-6f17d22bba15001f/fortanix-sgx-abi-0.5.0/src/lib.rs:873:43
    |
    |
873 |           { define_invoke_with_usercalls!(@ [$($accumulated)* $(fn $f($($n: $t),*) $(-> $r)*;)*] $($remainder)*); };
874 | |     // visit modules
874 | |     // visit modules
875 | |     (@ $accumulated:tt $(#[$meta:meta])* pub mod $modname:ident { $($contents:tt)* } $($remainder:tt)*) =>
876 | |         { define_invoke_with_usercalls!(@ $accumulated $($contents)* $($remainder)*); };
892 | |         macro_rules! invoke_with_usercalls {
892 | |         macro_rules! invoke_with_usercalls {
    | |         ---------------------------------- in this expansion of `invoke_with_usercalls!` (#1)
893 | |             ($m:ident) => { $m! $accumulated; }


error: type `fortanix_sgx_abi::Cancel` from private dependency 'fortanix_sgx_abi' in public interface
   --> library/std/src/sys/sgx/abi/usercalls/raw.rs:211:9
60  | / macro_rules! define_usercalls {
60  | / macro_rules! define_usercalls {
61  | |     ($(fn $f:ident($($n:ident: $t:ty),*) $(-> $r:tt)*; )*) => {
62  | |         /// Usercall numbers as per the ABI.
63  | |         #[repr(u64)]
...   |
74  | |         $(enclave_usercalls_internal_define_usercalls!(def fn $f($($n: $t),*) $(-> $r)*);)*
75  | |     };
76  | | }
    | |_- in this expansion of `define_usercalls!` (#2)
...
...
188 | / macro_rules! enclave_usercalls_internal_define_usercalls {
189 | |     (def fn $f:ident($n1:ident: $t1:ty, $n2:ident: $t2:ty,
190 | |                      $n3:ident: $t3:ty, $n4:ident: $t4:ty) -> $r:tt) => (
191 | |         /// This is the raw function definition, see the ABI documentation for
...   |
211 | |         pub unsafe fn $f($n1: $t1, $n2: $t2, $n3: $t3) -> $r {
...   |
266 | |     );
267 | | }
267 | | }
    | |_- in this expansion of `enclave_usercalls_internal_define_usercalls!` (#3)
268 |
269 |   invoke_with_usercalls!(define_usercalls);
    |
   ::: /cargo/registry/src/index.crates.io-6f17d22bba15001f/fortanix-sgx-abi-0.5.0/src/lib.rs:873:43
    |
    |
873 |           { define_invoke_with_usercalls!(@ [$($accumulated)* $(fn $f($($n: $t),*) $(-> $r)*;)*] $($remainder)*); };
874 | |     // visit modules
874 | |     // visit modules
875 | |     (@ $accumulated:tt $(#[$meta:meta])* pub mod $modname:ident { $($contents:tt)* } $($remainder:tt)*) =>
876 | |         { define_invoke_with_usercalls!(@ $accumulated $($contents)* $($remainder)*); };
892 | |         macro_rules! invoke_with_usercalls {
892 | |         macro_rules! invoke_with_usercalls {
    | |         ---------------------------------- in this expansion of `invoke_with_usercalls!` (#1)
893 | |             ($m:ident) => { $m! $accumulated; }

[RUSTC-TIMING] std test:false 2.292
warning: `std` (lib) generated 1 warning
error: could not compile `std` (lib) due to 11 previous errors; 1 warning emitted
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5028886449/jobs/9020290904)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/5028886449/jobs/9020290904","builder_name":"checks-actions"} -->

bors commented:
:pushpin: Commit 6c7b6a2d5afec1ca459a642d7efcc7e903618b18 has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 6c7b6a2d5afec1ca459a642d7efcc7e903618b18 -->
<!-- homu: {"type":"Approved","sha":"6c7b6a2d5afec1ca459a642d7efcc7e903618b18","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 6c7b6a2d5afec1ca459a642d7efcc7e903618b18 with merge b26f093e804b4068361298e86d195666a363dc9e...
<!-- homu: {"type":"BuildStarted","head_sha":"6c7b6a2d5afec1ca459a642d7efcc7e903618b18","merge_sha":"b26f093e804b4068361298e86d195666a363dc9e"} -->

rust-log-analyzer commented:
The job **`dist-various-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/5034935134/jobs/9030080358) [(plain)](https://github.com/rust-lang-ci/rust/commit/b26f093e804b4068361298e86d195666a363dc9e/checks/13633988539/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] gimli test:false 4.718
[RUSTC-TIMING] object test:false 5.098
warning: dropping unsupported crate type `dylib` for target `x86_64-fortanix-unknown-sgx`

error[E0364]: `accept_stream` is only public within the crate, and cannot be re-exported outside
  --> library/std/src/os/fortanix_sgx/mod.rs:25:13
   |
25 |             accept_stream, alloc, async_queues, bind_stream, close, connect_stream, exit, flush,
   |
   |
note: consider marking `accept_stream` as `pub` in the imported module
  --> library/std/src/os/fortanix_sgx/mod.rs:25:13
   |
25 |             accept_stream, alloc, async_queues, bind_stream, close, connect_stream, exit, flush,


error[E0364]: `alloc` is only public within the crate, and cannot be re-exported outside
  --> library/std/src/os/fortanix_sgx/mod.rs:25:28
   |
25 |             accept_stream, alloc, async_queues, bind_stream, close, connect_stream, exit, flush,
   |
   |
note: consider marking `alloc` as `pub` in the imported module
  --> library/std/src/os/fortanix_sgx/mod.rs:25:28
   |
25 |             accept_stream, alloc, async_queues, bind_stream, close, connect_stream, exit, flush,


error[E0364]: `async_queues` is only public within the crate, and cannot be re-exported outside
  --> library/std/src/os/fortanix_sgx/mod.rs:25:35
   |
25 |             accept_stream, alloc, async_queues, bind_stream, close, connect_stream, exit, flush,
   |
   |
note: consider marking `async_queues` as `pub` in the imported module
  --> library/std/src/os/fortanix_sgx/mod.rs:25:35
   |
25 |             accept_stream, alloc, async_queues, bind_stream, close, connect_stream, exit, flush,


error[E0364]: `bind_stream` is only public within the crate, and cannot be re-exported outside
  --> library/std/src/os/fortanix_sgx/mod.rs:25:49
   |
25 |             accept_stream, alloc, async_queues, bind_stream, close, connect_stream, exit, flush,
   |
   |
note: consider marking `bind_stream` as `pub` in the imported module
  --> library/std/src/os/fortanix_sgx/mod.rs:25:49
   |
25 |             accept_stream, alloc, async_queues, bind_stream, close, connect_stream, exit, flush,


error[E0364]: `close` is only public within the crate, and cannot be re-exported outside
  --> library/std/src/os/fortanix_sgx/mod.rs:25:62
   |
25 |             accept_stream, alloc, async_queues, bind_stream, close, connect_stream, exit, flush,
   |
   |
note: consider marking `close` as `pub` in the imported module
  --> library/std/src/os/fortanix_sgx/mod.rs:25:62
   |
25 |             accept_stream, alloc, async_queues, bind_stream, close, connect_stream, exit, flush,


error[E0364]: `connect_stream` is only public within the crate, and cannot be re-exported outside
  --> library/std/src/os/fortanix_sgx/mod.rs:25:69
   |
25 |             accept_stream, alloc, async_queues, bind_stream, close, connect_stream, exit, flush,
   |
   |
note: consider marking `connect_stream` as `pub` in the imported module
  --> library/std/src/os/fortanix_sgx/mod.rs:25:69
   |
25 |             accept_stream, alloc, async_queues, bind_stream, close, connect_stream, exit, flush,


error[E0364]: `exit` is only public within the crate, and cannot be re-exported outside
  --> library/std/src/os/fortanix_sgx/mod.rs:25:85
   |
25 |             accept_stream, alloc, async_queues, bind_stream, close, connect_stream, exit, flush,
   |
   |
note: consider marking `exit` as `pub` in the imported module
  --> library/std/src/os/fortanix_sgx/mod.rs:25:85
   |
25 |             accept_stream, alloc, async_queues, bind_stream, close, connect_stream, exit, flush,


error[E0364]: `flush` is only public within the crate, and cannot be re-exported outside
  --> library/std/src/os/fortanix_sgx/mod.rs:25:91
   |
25 |             accept_stream, alloc, async_queues, bind_stream, close, connect_stream, exit, flush,
   |
   |
note: consider marking `flush` as `pub` in the imported module
  --> library/std/src/os/fortanix_sgx/mod.rs:25:91
   |
25 |             accept_stream, alloc, async_queues, bind_stream, close, connect_stream, exit, flush,


error[E0364]: `free` is only public within the crate, and cannot be re-exported outside
  --> library/std/src/os/fortanix_sgx/mod.rs:26:13
   |
26 |             free, insecure_time, launch_thread, read, read_alloc, send, wait, write,
   |
   |
note: consider marking `free` as `pub` in the imported module
  --> library/std/src/os/fortanix_sgx/mod.rs:26:13
   |
26 |             free, insecure_time, launch_thread, read, read_alloc, send, wait, write,


error[E0364]: `insecure_time` is only public within the crate, and cannot be re-exported outside
  --> library/std/src/os/fortanix_sgx/mod.rs:26:19
   |
26 |             free, insecure_time, launch_thread, read, read_alloc, send, wait, write,
   |
   |
note: consider marking `insecure_time` as `pub` in the imported module
  --> library/std/src/os/fortanix_sgx/mod.rs:26:19
   |
26 |             free, insecure_time, launch_thread, read, read_alloc, send, wait, write,


error[E0364]: `launch_thread` is only public within the crate, and cannot be re-exported outside
  --> library/std/src/os/fortanix_sgx/mod.rs:26:34
   |
26 |             free, insecure_time, launch_thread, read, read_alloc, send, wait, write,
   |
   |
note: consider marking `launch_thread` as `pub` in the imported module
  --> library/std/src/os/fortanix_sgx/mod.rs:26:34
   |
26 |             free, insecure_time, launch_thread, read, read_alloc, send, wait, write,


error[E0364]: `read` is only public within the crate, and cannot be re-exported outside
  --> library/std/src/os/fortanix_sgx/mod.rs:26:49
   |
26 |             free, insecure_time, launch_thread, read, read_alloc, send, wait, write,
   |
   |
note: consider marking `read` as `pub` in the imported module
  --> library/std/src/os/fortanix_sgx/mod.rs:26:49
   |
26 |             free, insecure_time, launch_thread, read, read_alloc, send, wait, write,


error[E0364]: `read_alloc` is only public within the crate, and cannot be re-exported outside
  --> library/std/src/os/fortanix_sgx/mod.rs:26:55
   |
26 |             free, insecure_time, launch_thread, read, read_alloc, send, wait, write,
   |
   |
note: consider marking `read_alloc` as `pub` in the imported module
  --> library/std/src/os/fortanix_sgx/mod.rs:26:55
   |
26 |             free, insecure_time, launch_thread, read, read_alloc, send, wait, write,


error[E0364]: `send` is only public within the crate, and cannot be re-exported outside
  --> library/std/src/os/fortanix_sgx/mod.rs:26:67
   |
26 |             free, insecure_time, launch_thread, read, read_alloc, send, wait, write,
   |
   |
note: consider marking `send` as `pub` in the imported module
  --> library/std/src/os/fortanix_sgx/mod.rs:26:67
   |
26 |             free, insecure_time, launch_thread, read, read_alloc, send, wait, write,


error[E0364]: `wait` is only public within the crate, and cannot be re-exported outside
  --> library/std/src/os/fortanix_sgx/mod.rs:26:73
   |
26 |             free, insecure_time, launch_thread, read, read_alloc, send, wait, write,
   |
   |
note: consider marking `wait` as `pub` in the imported module
  --> library/std/src/os/fortanix_sgx/mod.rs:26:73
   |
26 |             free, insecure_time, launch_thread, read, read_alloc, send, wait, write,


error[E0364]: `write` is only public within the crate, and cannot be re-exported outside
  --> library/std/src/os/fortanix_sgx/mod.rs:26:79
   |
26 |             free, insecure_time, launch_thread, read, read_alloc, send, wait, write,
   |
   |
note: consider marking `write` as `pub` in the imported module
  --> library/std/src/os/fortanix_sgx/mod.rs:26:79
   |
26 |             free, insecure_time, launch_thread, read, read_alloc, send, wait, write,


error[E0364]: `do_usercall` is only public within the crate, and cannot be re-exported outside
  --> library/std/src/os/fortanix_sgx/mod.rs:28:51
   |
28 |         pub use crate::sys::abi::usercalls::raw::{do_usercall, Usercalls as UsercallNrs};
   |
   |
note: consider marking `do_usercall` as `pub` in the imported module
  --> library/std/src/os/fortanix_sgx/mod.rs:28:51
   |
28 |         pub use crate::sys::abi::usercalls::raw::{do_usercall, Usercalls as UsercallNrs};


error[E0364]: `free` is only public within the crate, and cannot be re-exported outside
   --> library/std/src/sys/sgx/abi/usercalls/mod.rs:265:9
265 | pub use self::raw::free;
    |         ^^^^^^^^^^^^^^^
    |
    |
note: consider marking `free` as `pub` in the imported module
   --> library/std/src/sys/sgx/abi/usercalls/mod.rs:265:9
265 | pub use self::raw::free;
    |         ^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0364`.
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5034935134/jobs/9030311772)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/5034935134/jobs/9030311772","builder_name":"checks-actions"} -->

notriddle committed:
std: make `fortanix-sgx-abi` a public depedenceny

It's exported publicly, so it should not be linted.

bors commented:
:umbrella: The latest upstream changes (presumably #111925) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

cjgillot commented:
@bors r+

lukas-code commented on compiler/rustc_data_structures/src/sync.rs:
```diff
@@ -143,6 +143,11 @@ cfg_if! {
                 result
             }
+            pub fn fetch_and(&self, val: bool, _: Ordering) -> bool {
+                let result = self.0.get() & val;
+                self.0.set(val);
+                result
```
This currently does `self.swap(val) & val`, not `self.fetch_and(val)`. Same for the "`fetch_or`" above.

```suggestion
                let old = self.0.get();
                self.0.set(old & val);
                old
```

bors commented:
:hourglass: Testing commit 5fb752bdd51b2d40ffbddf3c86e5869b3e816e08 with merge ee838d0ea3588d5441eb64196c145322243b5d00...
<!-- homu: {"type":"BuildStarted","head_sha":"5fb752bdd51b2d40ffbddf3c86e5869b3e816e08","merge_sha":"ee838d0ea3588d5441eb64196c145322243b5d00"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-distcheck`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/5084497000/jobs/9136880846) [(plain)](https://github.com/rust-lang-ci/rust/commit/ee838d0ea3588d5441eb64196c145322243b5d00/checks/13769554329/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Distcheck rust-src
error: failed to parse manifest at `/checkout/obj/build/tmp/distcheck-src/rust-src/lib/rustlib/src/rust/library/std/Cargo.toml`

Caused by:
  the cargo feature `public-dependency` requires a nightly version of Cargo, but this is the `beta` channel
  See https://doc.rust-lang.org/book/appendix-07-nightly-rust.html for more information about Rust release channels.
  See https://doc.rust-lang.org/beta/cargo/reference/unstable.html#public-dependency for more information about using this feature.
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5084497000/jobs/9137960520)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/5084497000/jobs/9137960520","builder_name":"checks-actions"} -->

notriddle committed:
bootstrap: use RUSTC_BOOTSTRAP in distcheck

notriddle committed:
rustc_data_structures: sync and atomic consistency

Co-authored-by: @lukas-code

cjgillot commented:
@bors r+

bors commented:
:pushpin: Commit 52bd82f522c4f3d9bd0dc534c06169285afbc23b has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 52bd82f522c4f3d9bd0dc534c06169285afbc23b -->
<!-- homu: {"type":"Approved","sha":"52bd82f522c4f3d9bd0dc534c06169285afbc23b","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

rust-timer commented:
Finished benchmarking commit (ad8304a0d5280de30856b39c19df7b306957e878): [comparison URL](https://perf.rust-lang.org/compare.html?start=e4f7ad8e68206fa54372535b5c04d7ddbaea43be&end=ad8304a0d5280de30856b39c19df7b306957e878&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=e4f7ad8e68206fa54372535b5c04d7ddbaea43be&end=ad8304a0d5280de30856b39c19df7b306957e878&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=e4f7ad8e68206fa54372535b5c04d7ddbaea43be&end=ad8304a0d5280de30856b39c19df7b306957e878&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=e4f7ad8e68206fa54372535b5c04d7ddbaea43be&end=ad8304a0d5280de30856b39c19df7b306957e878&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=e4f7ad8e68206fa54372535b5c04d7ddbaea43be&end=ad8304a0d5280de30856b39c19df7b306957e878&stat=size:linked_artifact)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.0% | [0.0%, 0.0%] | 14    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

**Bootstrap**: 643.343s -> 642.658s (-0.11%)

danakj commented:
Hey, it looks like this PR breaks building the toolchain with the Cargo binaries specified in https://github.com/rust-lang/rust/blob/master/src/stage0.json#L28-L83

The public-dependencies feature is a nightly feature, but the Cargo given there is beta.

```
linux_upload_rust log:
 Downloading https://chrome-infra-packages.appspot.com/dl/infra/3pp/static_libs/openssl/linux-amd64/+/version:2@1.1.1j.chromium.2 .......... Done.
 Downloading https://static.rust-lang.org/dist/2023-05-30/cargo-beta-x86_64-unknown-linux-gnu.tar.gz .......... Done.
 Running git submodule update --init --recursive --depth 1
 Running /b/s/w/ir/cache/builder/src/third_party/llvm-build-tools/cargo-beta-x86_64-unknown-linux-gnu/cargo/bin/cargo vendor --locked --versioned-dirs --sync ./src/tools/cargo/Cargo.toml --sync ./src/tools/rust-analyzer/Cargo.toml --sync ./compiler/rustc_codegen_cranelift/Cargo.toml --sync ./src/bootstrap/Cargo.toml
 error: failed to load manifest for workspace member `/b/s/w/ir/cache/builder/src/third_party/rust_src/src/library/std`
 
 Caused by:
   failed to parse manifest at `/b/s/w/ir/cache/builder/src/third_party/rust_src/src/library/std/Cargo.toml`
 
 Caused by:
   the cargo feature `public-dependency` requires a nightly version of Cargo, but this is the `beta` channel
   See https://doc.rust-lang.org/book/appendix-07-nightly-rust.html for more information about Rust release channels.
   See https://doc.rust-lang.org/beta/cargo/reference/unstable.html#public-dependency for more information about using this feature.
 Failed.
 Failed cargo vendor, retrying...
 Running git submodule update --init --recursive --depth 1
 Running /b/s/w/ir/cache/builder/src/third_party/llvm-build-tools/cargo-beta-x86_64-unknown-linux-gnu/cargo/bin/cargo vendor --locked --versioned-dirs --sync ./src/tools/cargo/Cargo.toml --sync ./src/tools/rust-analyzer/Cargo.toml --sync ./compiler/rustc_codegen_cranelift/Cargo.toml --sync ./src/bootstrap/Cargo.toml
 error: failed to load manifest for workspace member `/b/s/w/ir/cache/builder/src/third_party/rust_src/src/library/std`
 
 Caused by:
   failed to parse manifest at `/b/s/w/ir/cache/builder/src/third_party/rust_src/src/library/std/Cargo.toml`
 
 Caused by:
   the cargo feature `public-dependency` requires a nightly version of Cargo, but this is the `beta` channel
   See https://doc.rust-lang.org/book/appendix-07-nightly-rust.html for more information about Rust release channels.
   See https://doc.rust-lang.org/beta/cargo/reference/unstable.html#public-dependency for more information about using this feature.
 Failed.
```

How do we continue building the toolchain in this state?

Chromium bug: https://bugs.chromium.org/p/chromium/issues/detail?id=1450894

bjorn3 commented:
I don't think `RUSTC_BOOTSTRAP` should be suggested. The cargo version used to vendor may not support `public-dependency` or it may have an incompatible implementation. By the way I opened https://github.com/rust-lang/rust/issues/112391 a couple of days ago as I didn't notice this discussion.
